### ISSUE TITLE
"internal" node properties in Fusion/UI

### ISSUE THREAD

**Issue by kitsunet** [neos/neos-development-collection#4208]
Currently there is this in Neos.Neos/NodeTypes/Mixin/Node.yaml

```
  properties:
    #TODO: IS IT BREAKING TO DISABLE THESE NODE PROPERTIES BELOW WITH ES CR?
    #_creationDateTime:
    #  type: DateTime
    #_lastModificationDateTime:
    #  type: DateTime
    #_lastPublicationDateTime:
    #  type: DateTime
    #_path:
    #  type: string
    #_name:
    #  type: string
```

And yes unsurprisingly this is breaking in multiple ways. These are assumed ot be there at least in the NodeInfoView but probably in other places, also userland code might use those ala `q(node).property('_path')` which is perfectly "legal" at the moment. So we need to adapt internal usage and create migrations for userland code if we want to get rid of those (as well as create an accessor for them I guess?)

**Comment by kitsunet** [comment]
This is related to #3434 as that is using them. However I have a fix prepared for that.

**Comment by mhsdesign** [comment]
in 7.3 for example these are the node properties which can be accessed via the internal property syntax:

properties which are declared in the original NodeInterface:
```
_accessRestrictions
_accessRoles
_accessible
_autoCreated
_childNodes
_contentObject
_context
_contextPath
_depth
_dimensions
_hidden
_hiddenAfterDateTime
_hiddenBeforeDateTime
_hiddenInIndex
_identifier
_index
_label
_name
_node
_nodeData
_nodeType
_nodeTypeAllowedAsChildNode
_otherNodeVariants
_parent
_parentPath
_path
_primaryChildNode
_properties
_property
_propertyNames
_removed
_visible
_workspace
```

for completion (unrelated but i just wanted to document it) these are the settable properties a la NodeInterface:
```
_accessRoles
_contentObject
_hidden
_hiddenAfterDateTime
_hiddenBeforeDateTime
_hiddenInIndex
_index
_name
_nodeType
_property
_removed
_workspace
```

additionally accessible properties because of the TraversableNodeInterface and Node.
```
_cacheEntryIdentifier
_contentStreamIdentifier
_creationDateTime
_dimensionSpacePoint
_lastModificationDateTime
_lastPublicationDateTime
_nodeAggregateIdentifier
_nodeName
_nodeTypeName
_numberOfChildNodes
_originDimensionSpacePoint
_root
_tethered
```



`_path` and `_identifier` are still supported in the new 9.0 cr:

https://github.com/neos/neos-development-collection/blob/8d7acb506d87c76b47bf9114dddd696bee15fdab/Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/PropertyOperation.php#L93C1-L100

but since we kept the dynamic internal access in 9.0

```php
if ($propertyName[0] === '_') {
    return ObjectAccess::getPropertyPath($element, substr($propertyName, 1));
}
```

we will also allow in the future some magic wizardry for these (new) 9.0 properties:

```
_subgraphIdentity
_nodeAggregateId
_originDimensionSpacePoint
_classification
_nodeTypeName
_nodeType
_properties
_nodeName
_timestamps
```

im against that. Keeping a legacy layer: Ok but i would refrain from allowing new magic accessors.

also there are futher cases like here: https://github.com/neos/neos-development-collection/blob/7ac994259f5ee637533dd1a9e1fd8516688b386d/Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/FilterOperation.php#L123

**Comment by kitsunet** [comment]
Suggestion mentioned in daily today is to actually fully abandon these as it will just lead to confusion with people when some still work. Lets rather migrate them and offer alternatives where it makes sense.

**Comment by mhsdesign** [comment]
100% agree.

Additionally we should discuss if `_hiddenInIndex` - which in 8.3 was a magic property and now in 9.0 is a normal but untypical property due to the underscore - should be renamed to `hiddenInIndex`. Otherwise we could say this is the last core property which requires an underscore.

**Comment by dlubitz** [comment]
Please check also the already created rector migrations:
https://github.com/neos/rector/pull/39

**Comment by mhsdesign** [comment]
To avoid confusion, and since we have rector migrations in place, we decided in the weekly that we wanted to drop the legacy layer: https://github.com/neos/neos-development-collection/blob/8d7acb506d87c76b47bf9114dddd696bee15fdab/Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/PropertyOperation.php#L93C1-L100 and not further support the underscore `_` syntax anymore.

We should especially look at some node types as the underscore `_` syntax is there still used for example to define an editor for the `_nodeType` but that should rather be either hardcoded in the ui or declared elsewhere. I think neos 9 doesnt need the concept of underscores for untypical properties anymore, and they are opened for consumer land to be used freely.

**Comment by kitsunet** [comment]
Unfortunately we actually added `_hidden` and `_hiddenInIndex` into the property declaration and use them as _regular_ properties now, they are just in the json like anything else. Seems a bit big break to change that still?

**Comment by dlubitz** [comment]
What about a NodeMigration to rename the property?

**Comment by kitsunet** [comment]
But rename it to what? 😆 The thing is ... complicated, I can explain in the weekly... or write it out here later.

**Comment by mhsdesign** [comment]
`_hidden` and `_nodeType` must go from the nodeType yaml imo.

**Comment by bwaidelich** [comment]
As decided in our weekly today, the plan is to:
* Rename `_hiddenInIndex` to `hiddenInMenu`
* Remove `_hidden`
* Replace all remaining underscore properties in the core
* Remove special handling for underscores in properties

**Comment by mhsdesign** [comment]
Todo: explicit eel ordering support via time stamp fields: https://neos-project.slack.com/archives/C050C8FEK/p1709887830143079?thread_ts=1709403062.357079&cid=C050C8FEK

currently in 9.0-dev its `sort('_timestamps.originalCreated')`

**Comment by mhsdesign** [comment]
Not yet -> https://github.com/neos/neos-development-collection/pull/5015

**Comment by mhsdesign** [comment]
Update 2024 these are all the properties accessible in the Neos 8 Node (via `ObjectAccess::getGettablePropertyNames`)

- `accessRestrictions`
- `accessRoles`
- `accessible`
- `autoCreated`
- `cacheEntryIdentifier`, not part of 8.3 API :ghost:
- `childNodes`
- `contentObject`
- `contentStreamIdentifier`, not part of 8.3 API :ghost:, **throws** in 8.3 NodeMethodIsUnsupported :boom:
- `context`, marked as **internal** in 8.3 :ghost:
- `contextPath`
- `creationDateTime`, not part of 8.3 API :ghost:
- `depth`
- `dimensionSpacePoint`, not part of 8.3 API :ghost:, available in **9.0** :white_check_mark:, **throws** in 8.3 NodeMethodIsUnsupported :boom:
- `dimensions`
- `hidden`
- `hiddenAfterDateTime`
- `hiddenBeforeDateTime`
- `hiddenInIndex`
- `identifier`
- `index`
- `label`
- `lastModificationDateTime`, not part of 8.3 API :ghost:
- `lastPublicationDateTime`, not part of 8.3 API :ghost:
- `name`, available in **9.0** :white_check_mark:
- `nodeAggregateIdentifier`, not part of 8.3 API :ghost:
- `nodeData`, marked as **internal** in 8.3 :ghost:
- `nodeName`, not part of 8.3 API :ghost:
- `nodeType`
- `nodeTypeName`, not part of 8.3 API :ghost:, available in **9.0** :white_check_mark:
- `numberOfChildNodes`, not part of 8.3 API :ghost:
- `originDimensionSpacePoint`, not part of 8.3 API :ghost:, available in **9.0** :white_check_mark:, **throws** in 8.3 NodeMethodIsUnsupported :boom:
- `otherNodeVariants`
- `parent`
- `parentPath`
- `path`
- `primaryChildNode`
- `properties`, available in **9.0** :white_check_mark:
- `propertyNames`
- `removed`
- `root`, not part of 8.3 API :ghost:
- `tethered`, not part of 8.3 API :ghost:
- `visible`
- `workspace`

**new additional properties in 9.0**

- `aggregateId`, only **9.0** :sparkles:
- `classification`, only **9.0** :sparkles:
- `contentRepositoryId`, only **9.0** :sparkles:
- `tags`, only **9.0** :sparkles:
- `timestamps`, only **9.0** :sparkles:
- `visibilityConstraints`, only **9.0** :sparkles:
- `workspaceName`, only **9.0** :sparkles:

**Invalid gettable properties as they require arguments:**
- `nodeTypeAllowedAsChildNode`, **throws** in 8.3 ArgumentCountError :boom:
- `property`, available in **9.0** :white_check_mark:, **throws** in 8.3 ArgumentCountError :boom:
- `node`, **throws** in 8.3 ArgumentCountError :boom:

**legend**

- not part of 8.3 API :ghost:, meaning not declared in the "original" node interface `\Neos\ContentRepository\Domain\Model\NodeInterface`
- marked as **internal** in 8.3 :ghost:, meaning it is explicitly declared as `@internal`
